src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -246,38 +239,81 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
-      }
-
-      if (!spec.GetDependencies().empty()) {
-        task_dependency_manager_.RemoveTaskDependencies(
-            task.GetTaskSpecification().TaskId());
+        work->status = WorkStatus::WAITING_FOR_WORKER;
+        worker_pool_.PopWorker(
+            spec,
+            [this, allocated_instances, task_id, scheduling_class, work](
+                const std::shared_ptr<WorkerInterface> worker, Status status) -> bool {
+              const auto &reply = work->reply;
+              const auto &callback = work->callback;
+              bool canceled = work->status == WorkStatus::CANCELLED;
+              const auto &task = work->task;
+              const auto &spec = task.GetTaskSpecification();
+              bool used = false;
+              auto erase_from_dispatch_queue_fn =
+                  [this](const std::shared_ptr<Work> &work,
+                         const SchedulingClass &scheduling_class) {
+                    auto shapes_it = tasks_to_dispatch_.find(scheduling_class);
+                    RAY_CHECK(shapes_it != tasks_to_dispatch_.end());
+                    auto &dispatch_queue = shapes_it->second;
+                    bool erased = false;
+                    for (auto work_it = dispatch_queue.begin();
+                         work_it != dispatch_queue.end(); work_it++) {
+                      if (*work_it == work) {
+                        dispatch_queue.erase(work_it);
+                        erased = true;
+                        break;
+                      }
+                    }
+                    if (dispatch_queue.empty()) {
+                      tasks_to_dispatch_.erase(shapes_it);
+                    }
+                    RAY_CHECK(erased);
+                  };
+
+              if (canceled || !worker || status.IsOwnerFailed()) {
+                if (canceled) {
+                  RAY_LOG(DEBUG)
+                      << "Task " << task_id
+                      << " has been canceled when worker popped, maybe the task has "
+                         "been cancelled.";
+                } else {
+                  RAY_LOG(DEBUG)
+                      << "This node has available resources, but no worker processes "
+                         "to grant the lease "
+                      << task_id << ", status = " << status.CodeAsString();
+                }
+                // We've already acquired resources so we need to release them.
+                cluster_resource_scheduler_->ReleaseWorkerResources(allocated_instances);
+                // Release pinned task args.
+                ReleaseTaskArgs(task_id);
+                if (status.IsOwnerFailed() && !canceled) {
+                  erase_from_dispatch_queue_fn(work, scheduling_class);
+
+                } else if (status.IsRuntimeEnvCreationFailed()) {
+                  CancelTask(task_id, true);
+                } else {
+                  work->status = WorkStatus::WAITING;
+                }
```

Comment:
Can you add some comments here to explain why the different cases are handled this way?

Reply:
Also, is this behavior the same as the previous code? I didn't think that we canceled tasks if the runtime env creation failed, but not sure.

Reply:
Sure, I added some comments.

Reply:
The runtime env failed behavior has been merged in this PR. https://github.com/ray-project/ray/commit/7007c6271d4becb762ab33d65a6841824a15134e